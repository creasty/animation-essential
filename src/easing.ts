export type EasingFunc = (t: number) => number;

const BACK = {
  strength: 1.525,
};

const BOUNCE = {
  t1: 4 / 11,
  t2: 8 / 11,
  t3: 9 / 10,
  a: 4356 / 361,
  b: 35442 / 1805,
  c: 16061 / 1805,
};

const { sin, cos, PI, sqrt, pow } = Math;

export const Easing = Object.freeze({
  Linear: (t: number) => t,
  Quad: buildTriple((t) => t * t),
  Cubic: buildTriple((t) => t * t * t),
  Quart: buildTriple((t) => t * t * t * t),
  Quint: buildTriple((t) => t * t * t * t * t),
  Sine: buildTriple((t) => 1 - cos((t * PI) / 2)),
  Expo: buildTriple((t) => (t === 0 ? 0 : pow(1024, t - 1))),
  Circ: buildTriple((t) => 1 - sqrt(1 - t * t)),
  Elastic: buildTriple((t) => (t === 0 ? 0 : -pow(1024, t - 1) * sin((t - 1.1) * 5 * PI))),
  Back: buildTriple((t) => t * t * ((BACK.strength + 1) * t - BACK.strength)),
  Bounce: buildTriple((t) => {
    if (t === 1 || t === 0) return t;
    return t < BOUNCE.t1
      ? 7.5625 * t * t
      : t < BOUNCE.t2
      ? 9.075 * t * t - 9.9 * t + 3.4
      : t < BOUNCE.t3
      ? BOUNCE.a * t * t - BOUNCE.b * t + BOUNCE.c
      : 10.8 * t * t - 20.52 * t + 10.72;
  }, true),
});

export function reversedEasing(easing: EasingFunc): EasingFunc {
  return (t) => 1 - easing(1 - t);
}

export function mirroredEasing(easing: EasingFunc): EasingFunc {
  return (t) => (t < 0.5 ? easing(2 * t) * 0.5 : 1 - easing(2 * (1 - t)) * 0.5);
}

function buildTriple(easing: EasingFunc, reversed = false) {
  const [easeIn, easeOut] = reversed ? [reversedEasing(easing), easing] : [easing, reversedEasing(easing)];
  return {
    In: easeIn,
    Out: easeOut,
    InOut: mirroredEasing(easeIn),
  };
}
