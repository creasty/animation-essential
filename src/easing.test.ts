import { Easing, mirroredEasing, reversedEasing } from "./easing";

const ts = [0.0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1.0];

describe("reversedEasing", () => {
  test("returns a reversed function", () => {
    const fn = reversedEasing((t) => t * 0.5);
    expect(ts.map(fn)).toMatchInlineSnapshot(`
      [
        0.5,
        0.55,
        0.6,
        0.65,
        0.7,
        0.75,
        0.8,
        0.85,
        0.9,
        0.95,
        1,
      ]
    `);
  });
});

describe("mirroredEasing", () => {
  test("returns a mirrored function", () => {
    const fn = mirroredEasing((t) => t * 0.5);
    expect(ts.map(fn)).toMatchInlineSnapshot(`
      [
        0,
        0.05,
        0.1,
        0.15,
        0.2,
        0.75,
        0.8,
        0.85,
        0.9,
        0.95,
        1,
      ]
    `);
  });
});

describe("Easing", () => {
  const nonLinearFunctions: [Exclude<keyof typeof Easing, "Linear">, "In" | "Out" | "InOut"][] = [
    "Quad",
    "Cubic",
    "Quart",
    "Quint",
    "Sine",
    "Expo",
    "Circ",
    "Elastic",
    "Back",
    "Bounce",
  ].flatMap((name) => ["In", "Out", "InOut"].map((curveType) => [name, curveType])) as any;

  test("Linear being an identity function", () => {
    expect(ts.map(Easing.Linear)).toEqual(ts);
  });

  test.each(nonLinearFunctions)("%s.%s works", (name, curveType) => {
    const fn = Easing[name][curveType];
    expect(ts.map(fn)).toMatchSnapshot();
  });
});
