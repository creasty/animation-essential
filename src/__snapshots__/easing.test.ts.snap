// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Easing Back.In works 1`] = `
[
  -0,
  -0.012725000000000002,
  -0.04080000000000001,
  -0.069075,
  -0.0824,
  -0.06562499999999999,
  -0.003600000000000003,
  0.11882499999999996,
  0.31680000000000014,
  0.6054750000000001,
  1,
]
`;

exports[`Easing Back.InOut works 1`] = `
[
  -0,
  -0.020400000000000005,
  -0.0412,
  -0.0018000000000000015,
  0.15840000000000007,
  0.5,
  0.8415999999999999,
  1.0018,
  1.0412,
  1.0204,
  1,
]
`;

exports[`Easing Back.Out works 1`] = `
[
  0,
  0.3945249999999999,
  0.6831999999999998,
  0.881175,
  1.0036,
  1.065625,
  1.0824,
  1.069075,
  1.0408,
  1.012725,
  1,
]
`;

exports[`Easing Bounce.In works 1`] = `
[
  0,
  -1.7763568394002505e-15,
  0.08775623268697963,
  0.08325000000000093,
  0.273000000000001,
  0.28125000000000044,
  0.10800000000000054,
  0.31937499999999985,
  0.6975000000000001,
  0.9243750000000001,
  1,
]
`;

exports[`Easing Bounce.InOut works 1`] = `
[
  0,
  0.043878116343489815,
  0.1365000000000005,
  0.05400000000000027,
  0.34875000000000006,
  0.5,
  0.6512499999999999,
  0.946,
  0.8634999999999997,
  0.9561218836565102,
  1,
]
`;

exports[`Easing Bounce.Out works 1`] = `
[
  0,
  0.07562500000000001,
  0.30250000000000005,
  0.6806249999999999,
  0.8919999999999995,
  0.7187499999999996,
  0.726999999999999,
  0.9167499999999991,
  0.9122437673130204,
  1.0000000000000018,
  1,
]
`;

exports[`Easing Circ.In works 1`] = `
[
  0,
  0.005012562893380035,
  0.020204102886728803,
  0.04606079858305434,
  0.08348486100883201,
  0.1339745962155614,
  0.19999999999999996,
  0.285857157145715,
  0.40000000000000013,
  0.5641101056459328,
  1,
]
`;

exports[`Easing Circ.InOut works 1`] = `
[
  0,
  0.010102051443364402,
  0.041742430504416006,
  0.09999999999999998,
  0.20000000000000007,
  0.5,
  0.7999999999999999,
  0.8999999999999999,
  0.958257569495584,
  0.9898979485566356,
  1,
]
`;

exports[`Easing Circ.Out works 1`] = `
[
  0,
  0.4358898943540672,
  0.5999999999999999,
  0.714142842854285,
  0.8,
  0.8660254037844386,
  0.916515138991168,
  0.9539392014169457,
  0.9797958971132712,
  0.99498743710662,
  1,
]
`;

exports[`Easing Cubic.In works 1`] = `
[
  0,
  0.0010000000000000002,
  0.008000000000000002,
  0.027,
  0.06400000000000002,
  0.125,
  0.216,
  0.3429999999999999,
  0.5120000000000001,
  0.7290000000000001,
  1,
]
`;

exports[`Easing Cubic.InOut works 1`] = `
[
  0,
  0.004000000000000001,
  0.03200000000000001,
  0.108,
  0.25600000000000006,
  0.5,
  0.744,
  0.8919999999999999,
  0.968,
  0.996,
  1,
]
`;

exports[`Easing Cubic.Out works 1`] = `
[
  0,
  0.2709999999999999,
  0.4879999999999999,
  0.657,
  0.784,
  0.875,
  0.9359999999999999,
  0.973,
  0.992,
  0.999,
  1,
]
`;

exports[`Easing Elastic.In works 1`] = `
[
  0,
  1.195944139792337e-18,
  0.0039062499999999987,
  -3.8270212473354795e-18,
  -0.015625000000000003,
  -4.403008748925139e-17,
  0.06249999999999999,
  3.0245073740886303e-16,
  -0.25000000000000006,
  -6.0490147481772635e-16,
  1,
]
`;

exports[`Easing Elastic.InOut works 1`] = `
[
  0,
  0.0019531249999999993,
  -0.007812500000000002,
  0.031249999999999997,
  -0.12500000000000003,
  0.5,
  1.125,
  0.96875,
  1.0078125,
  0.998046875,
  1,
]
`;

exports[`Easing Elastic.Out works 1`] = `
[
  0,
  1.0000000000000007,
  1.25,
  0.9999999999999997,
  0.9375,
  1,
  1.015625,
  1,
  0.99609375,
  1,
  1,
]
`;

exports[`Easing Expo.In works 1`] = `
[
  0,
  0.0019531249999999998,
  0.0039062499999999987,
  0.007812500000000002,
  0.015625000000000003,
  0.03125,
  0.06249999999999999,
  0.12499999999999996,
  0.25000000000000006,
  0.5000000000000001,
  1,
]
`;

exports[`Easing Expo.InOut works 1`] = `
[
  0,
  0.0019531249999999993,
  0.007812500000000002,
  0.031249999999999997,
  0.12500000000000003,
  0.5,
  0.875,
  0.96875,
  0.9921875,
  0.998046875,
  1,
]
`;

exports[`Easing Expo.Out works 1`] = `
[
  0,
  0.4999999999999999,
  0.75,
  0.875,
  0.9375,
  0.96875,
  0.984375,
  0.9921875,
  0.99609375,
  0.998046875,
  1,
]
`;

exports[`Easing Quad.In works 1`] = `
[
  0,
  0.010000000000000002,
  0.04000000000000001,
  0.09,
  0.16000000000000003,
  0.25,
  0.36,
  0.48999999999999994,
  0.6400000000000001,
  0.81,
  1,
]
`;

exports[`Easing Quad.InOut works 1`] = `
[
  0,
  0.020000000000000004,
  0.08000000000000002,
  0.18,
  0.32000000000000006,
  0.5,
  0.6799999999999999,
  0.82,
  0.92,
  0.98,
  1,
]
`;

exports[`Easing Quad.Out works 1`] = `
[
  0,
  0.18999999999999995,
  0.3599999999999999,
  0.51,
  0.64,
  0.75,
  0.84,
  0.9099999999999999,
  0.96,
  0.99,
  1,
]
`;

exports[`Easing Quart.In works 1`] = `
[
  0,
  0.00010000000000000003,
  0.0016000000000000005,
  0.0081,
  0.025600000000000008,
  0.0625,
  0.1296,
  0.24009999999999992,
  0.40960000000000013,
  0.6561000000000001,
  1,
]
`;

exports[`Easing Quart.InOut works 1`] = `
[
  0,
  0.0008000000000000003,
  0.012800000000000004,
  0.0648,
  0.20480000000000007,
  0.5,
  0.7951999999999999,
  0.9351999999999999,
  0.9872,
  0.9992,
  1,
]
`;

exports[`Easing Quart.Out works 1`] = `
[
  0,
  0.3438999999999999,
  0.5903999999999998,
  0.7599,
  0.8704000000000001,
  0.9375,
  0.9744,
  0.9919,
  0.9984,
  0.9999,
  1,
]
`;

exports[`Easing Quint.In works 1`] = `
[
  0,
  0.000010000000000000004,
  0.00032000000000000013,
  0.00243,
  0.010240000000000004,
  0.03125,
  0.07776,
  0.16806999999999994,
  0.32768000000000014,
  0.5904900000000002,
  1,
]
`;

exports[`Easing Quint.InOut works 1`] = `
[
  0,
  0.00016000000000000007,
  0.005120000000000002,
  0.03888,
  0.16384000000000007,
  0.5,
  0.8361599999999999,
  0.96112,
  0.99488,
  0.99984,
  1,
]
`;

exports[`Easing Quint.Out works 1`] = `
[
  0,
  0.4095099999999998,
  0.6723199999999998,
  0.8319300000000001,
  0.92224,
  0.96875,
  0.98976,
  0.99757,
  0.99968,
  0.99999,
  1,
]
`;

exports[`Easing Sine.In works 1`] = `
[
  0,
  0.01231165940486223,
  0.04894348370484647,
  0.1089934758116321,
  0.19098300562505255,
  0.2928932188134524,
  0.41221474770752686,
  0.5460095002604531,
  0.6909830056250525,
  0.843565534959769,
  0.9999999999999999,
]
`;

exports[`Easing Sine.InOut works 1`] = `
[
  0,
  0.024471741852423234,
  0.09549150281252627,
  0.20610737385376343,
  0.3454915028125263,
  0.5,
  0.6545084971874737,
  0.7938926261462365,
  0.9045084971874737,
  0.9755282581475768,
  1,
]
`;

exports[`Easing Sine.Out works 1`] = `
[
  1.1102230246251565e-16,
  0.15643446504023095,
  0.30901699437494745,
  0.45399049973954686,
  0.5877852522924731,
  0.7071067811865476,
  0.8090169943749475,
  0.8910065241883679,
  0.9510565162951536,
  0.9876883405951378,
  1,
]
`;
